File: 03-provider-registry/golang/tools.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strings"
)

// ── Tool types ───────────────────────────────────────────────────────────────

type ToolParam struct {
	Type        string
	Description string
}

type ToolDefinition struct {
	Name        string
	Description string
	Properties  map[string]ToolParam
	Required    []string
}

type ToolExecutor func(params map[string]string) (string, error)

type Tool struct {
	Definition ToolDefinition
	Execute    ToolExecutor
}

// ── Tool registry ────────────────────────────────────────────────────────────

var toolRegistry = map[string]Tool{}

func registerTool(def ToolDefinition, execute ToolExecutor) {
	toolRegistry[def.Name] = Tool{Definition: def, Execute: execute}
}

func buildToolsPrompt() string {
	var sb strings.Builder
	first := true
	for _, tool := range toolRegistry {
		if !first {
			sb.WriteString("\n\n")
		}
		first = false
		d := tool.Definition
		fmt.Fprintf(&sb, "### %s\n%s\nParameters:", d.Name, d.Description)
		for k, v := range d.Properties {
			fmt.Fprintf(&sb, "\n  - %s (%s): %s", k, v.Type, v.Description)
		}
	}
	return sb.String()
}

func init() {
	registerTool(ToolDefinition{
		Name:        "shell",
		Description: "Execute a bash shell command and return stdout",
		Properties: map[string]ToolParam{
			"command": {Type: "string", Description: "The bash command to execute"},
		},
		Required: []string{"command"},
	}, func(params map[string]string) (string, error) {
		return runCommand(params["command"])
	})

	registerTool(ToolDefinition{
		Name:        "read_file",
		Description: "Read the content of a file",
		Properties: map[string]ToolParam{
			"path": {Type: "string", Description: "Absolute or relative file path"},
		},
		Required: []string{"path"},
	}, func(params map[string]string) (string, error) {
		data, err := os.ReadFile(params["path"])
		return string(data), err
	})
}

func runCommand(cmd string) (string, error) {
	var c *exec.Cmd
	if runtime.GOOS == "windows" {
		c = exec.Command("cmd", "/C", cmd)
	} else {
		c = exec.Command("sh", "-c", cmd)
	}
	out, err := c.CombinedOutput()
	return string(out), err
}

// ── JSON parsing helpers ─────────────────────────────────────────────────────

var (
	reJSONBlock     = regexp.MustCompile("(?s)```json\\s*(.*?)```")
	reRawBlock      = regexp.MustCompile("(?s)```\\s*(.*?)```")
	reInlineJSON    = regexp.MustCompile(`(?s)\{.*\}`)
	reInvalidEscape = regexp.MustCompile(`\\([^"\\/bfnrtu0-9])`)
)

func repairJSON(s string) string {
	return reInvalidEscape.ReplaceAllString(s, `\\$1`)
}

func tryParse(candidate string) map[string]any {
	candidate = strings.TrimSpace(candidate)
	var result map[string]any
	if json.Unmarshal([]byte(candidate), &result) == nil {
		return result
	}
	if json.Unmarshal([]byte(repairJSON(candidate)), &result) == nil {
		return result
	}
	return nil
}

func extractJSON(text string) map[string]any {
	s := strings.TrimSpace(text)

	if r := tryParse(s); r != nil {
		return r
	}
	if m := reJSONBlock.FindStringSubmatch(s); m != nil {
		if r := tryParse(m[1]); r != nil {
			return r
		}
	}
	if m := reRawBlock.FindStringSubmatch(s); m != nil {
		if r := tryParse(m[1]); r != nil {
			return r
		}
	}
	if m := reInlineJSON.FindString(s); m != "" {
		if r := tryParse(m); r != nil {
			return r
		}
	}
	return nil
}

// ── Helpers ──────────────────────────────────────────────────────────────────

func buildProviderChain() []string {
	primary := os.Getenv("PRIMARY_PROVIDER")
	if primary == "" {
		primary = "claude"
	}
	fallback := os.Getenv("FALLBACK_PROVIDER")
	if fallback == "" {
		fallback = "openai"
	}
	if primary == fallback {
		return []string{primary}
	}
	return []string{primary, fallback}
}

func buildSystemPrompt() string {
	return `You are an AI assistant named xclaw.

To use a tool, output a JSON object (bare or in a markdown code block):
{"action": "<tool_name>", "<param1>": "<value1>", ...}

To answer directly, output plain text — do NOT use JSON.

Available tools:
` + buildToolsPrompt()
}

func availableTools() string {
	keys := make([]string, 0, len(toolRegistry))
	for k := range toolRegistry {
		keys = append(keys, k)
	}
	return strings.Join(keys, ", ")
}

File: 04-realtime-communication/nodejs/src/tools.ts
import { execSync } from 'child_process';
import { readFileSync } from 'fs';

// ── Tool type definitions ────────────────────────────────────────────────────

interface ToolParam {
  type: string;
  description: string;
}

interface ToolDefinition {
  name: string;
  description: string;
  parameters: {
    type: 'object';
    properties: Record<string, ToolParam>;
    required: string[];
  };
}

type ToolExecutor = (params: Record<string, string>) => string;

interface Tool {
  definition: ToolDefinition;
  execute: ToolExecutor;
}

// ── Tool registry ────────────────────────────────────────────────────────────

export const toolRegistry = new Map<string, Tool>();

export function registerTool(definition: ToolDefinition, execute: ToolExecutor) {
  toolRegistry.set(definition.name, { definition, execute });
}

export function buildToolsPrompt(): string {
  return [...toolRegistry.values()]
    .map(({ definition: d }) => {
      const params = Object.entries(d.parameters.properties)
        .map(([k, v]) => `  - ${k} (${v.type}): ${v.description}`)
        .join('\n');
      return `### ${d.name}\n${d.description}\nParameters:\n${params}`;
    })
    .join('\n\n');
}

// ── Register tools ───────────────────────────────────────────────────────────

registerTool(
  {
    name: 'shell',
    description: 'Execute a bash shell command and return stdout',
    parameters: {
      type: 'object',
      properties: {
        command: { type: 'string', description: 'The bash command to execute' },
      },
      required: ['command'],
    },
  },
  ({ command }) => execSync(command, { encoding: 'utf-8' }),
);

registerTool(
  {
    name: 'read_file',
    description: 'Read the content of a file',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'Absolute or relative file path' },
      },
      required: ['path'],
    },
  },
  ({ path }) => readFileSync(path, 'utf-8'),
);

// ── JSON parsing helpers ─────────────────────────────────────────────────────

function repairJSON(s: string): string {
  return s.replace(/\\([^"\\/bfnrtu\d])/g, '\\\\$1');
}

function tryParse(candidate: string): Record<string, unknown> | null {
  try { return JSON.parse(candidate); } catch {}
  try { return JSON.parse(repairJSON(candidate)); } catch {}
  return null;
}

export function extractJSON(text: string): Record<string, unknown> | null {
  const s = text.trim();

  const r1 = tryParse(s);
  if (r1) return r1;

  const jsonBlock = s.match(/```json\s*([\s\S]*?)```/);
  if (jsonBlock) { const r = tryParse(jsonBlock[1].trim()); if (r) return r; }

  const rawBlock = s.match(/```\s*([\s\S]*?)```/);
  if (rawBlock) { const r = tryParse(rawBlock[1].trim()); if (r) return r; }

  const inlineMatch = s.match(/\{[\s\S]*\}/);
  if (inlineMatch) { const r = tryParse(inlineMatch[0]); if (r) return r; }

  return null;
}
